### PR TITLE
Add generic TimeDistributed layer.

### PR DESCRIPTION
**Pull request by fchollet**
This is what I had in mind initially. The same principle can also be applied to `Residual`, `Highway`, etc.

@farizrahman4u @lukedeo @EderSantana any feedback?

### CODE REVIEW

**Comment by fchollet** [comment]
@farizrahman4u just checked out your PR again, I see you went for a lot of the same stuff. I agree that the `Wrapper` abstraction is useful and should be used as soon as we have more than one wrapper. In general your PR misses tests, and should be broken up into smaller pieces (one wrapper at a time). Although it misses a few things, such as masking, model serialization / loading.

**Comment by farizrahman4u** [comment]
Looks great. And complete, when compared to my PR. Still, all the reshaping / building business could be abstracted into a base `Wrapper` class. 

P.S Not related to this, but we still haven't fixed the RNN go_backwards problem. Sequence and mask being out of alignment is a serious issue, please look into it.

**Comment by fchollet** [comment]
> Still, all the reshaping / building business could be abstracted into a base Wrapper class.

Completely agreed, pretty every wrapper will have these operations in common. Do you want to submit a first PR with just the `Wrapper` class?

**Comment by farizrahman4u** [comment]
#1828

**Comment by farizrahman4u** [comment]
One limitation of the current implementation is that the input shape has to be provided even if its not the first layer in the model.

```python
model = Sequential()
model.add(wrappers.TimeDistributed(core.Dense(2), input_shape=(3, 4)))
model.add(wrappers.TimeDistributed(core.Dense(10))  # Will not work as the layer is not built during __init__()
model.add(core.Activation('relu'))
```
#1828 Will help you handle this issue.

**Comment by EderSantana** [comment]
When I implemented TimeDistributed here: https://github.com/EderSantana/seya/blob/master/seya/layers/conv_rnn.py#L315-L328

I went with reshaping and no scan, I think this made compilation time lower. But on the other hand. It makes us require a batch size for it to work with tensorflow. But the solution here looks cleaner than mine.

 About the reshaping business, maybe we don't want to "force" it to work, since it may hide bugs in the layer design.

**Comment by farizrahman4u** [comment]
Drop the `rnn()` just like we did for `TimeDistributedDense`

```python
def get_output(self, train=False):
    X = self.get_input(train)  # (nb_samples, timesteps, ..., input_dim)
    input_shape = self.input_shape
    x = K.reshape(X, (-1, ) + input_shape[2:])  # (nb_samples * timesteps, ..., input_dim)
    Y = self.layer(x, train=False)
    input_length = input_shape[1]
    if not input_length:
        input_length = K.shape(X)[1]
    Y = K.reshape(Y, (-1, input_length) + layer.output_shape[1:])
    return Y
```

**Comment by fchollet** [comment]
@farizrahman4u it's dangerous, because it assumes the batch size doesn't matter. For the stateful case, you by-pass this with a special if condition (we should avoid special cases in general), but there are actually plenty of other situations in which the batch size matters. For my own research I have developed a collection of experimental layers that require a batch size...

I'd replace it with:

```python
if self.input_shape[0]:
   # batch size matters, use rnn-based implementation
else:
  # no batch size specified, therefore the layer will be able to process batches of any size
  # we can go with reshape-based implementation for performance
```

**Comment by fchollet** [comment]
Also a TF-specific exception should be raised if `timesteps` wasn't specified.

**Comment by fchollet** [comment]
> And mask can be ignored.

Why?

**Comment by farizrahman4u** [comment]
> Why?

Because there is no state involved. The same reason  why we e don't care for mask in `TimeDistribuedDense`.Consider a layer that does `f(x) = 2 * x` and we wrap it in a time distributed wrapper and process the following sequence + mask:

```
sequence: 1 2 3 4 5 6 7 8
mask: 1 1 1 1 0 0 0
``` 

The output would be:

```
sequence: 2 4 6 8 8 8 8 8
mask: 1 1 1 1 0 0 0
```

If you were to ignore the mask, the output would be:

```
sequence: 2 4 6 8 10 12 14 16
mask: 1 1 1 1 0 0 0
```

The unmasked elements are the same and we don't care what the masked elements are anyway.

**Comment by farizrahman4u** [comment]
So, this is what `get_output` looks like now. I am pasting code here instead PRing so that we can have the whole conversation in one place.

```python
def get_output(self, train=False):
    X = self.get_input(train)  # (nb_samples, timesteps, ..., input_dim)

    # Use rnn if the child requires a fixed batch size
    if self.layer.input_sahpe[0]:
        def step(x, states):
            return self.layer(x), []
        Y = K.rnn(step, X, [])[1]
        return Y
    
    input_shape = self.input_shape
    x = K.reshape(X, (-1, ) + input_shape[2:])  # (nb_samples * timesteps, ..., input_dim)
    Y = self.layer(x, train=False)  # (nb_samples * timesteps, ...., output_dim)
    input_length = input_shape[1]
    if not input_length:
        input_length = K.shape(X)[1]
    Y = K.reshape(Y, (-1, input_length) + self.layer.output_shape[1:])  # (nb_samples, timesteps, ..., output_dim)
    return Y
```

**Comment by farizrahman4u** [comment]
An issue of the top of my head:

```
model = Sequential()
model.add(LSTM(batch_input_shape=(7, 10, 5), output_dim=10, stateful=True, return_sequences=True))
model.add(TimeDistributed(Dense(10)))
```

Here, the input shape for the `Dense` layer is not set, so the wrapper would set the input shape for it, (#1828). So the input shape of the `Dense` layer would be `(7, 5)`. Then `TimeDistributed` would think that batch is important for the dense layer (`if self.layer.input_shape[0]`) and use the slow RNN implementation. Of course, can be fixed by setting the input shape of the dense layer explicitly.

**Comment by fchollet** [comment]
In this context `self.layer.input_shape` is the `TimeDistributed` input shape, not the input shape of the child layer. So in any case the batch size would be set and the `rnn` implementation would be used.

It's kind of an edge case.

**Comment by farizrahman4u** [comment]
> In this context self.layer.input_shape is the TimeDistributed input shape

Not exactly, the `timesteps` axis would be removed. Just saying, does not contradict with the point you are making. Edge case it is.
